indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java

```diff
@@ -298,6 +339,120 @@ public TaskLock lock(final Task task, final Interval interval) throws Interrupte
+    for (Map.Entry<String, TaskLock> entry : tasksToPreempt.entrySet()) {
+      Task task = taskStorage.getTask(entry.getKey()).get();
+      log.info(
+          "Revoking task lock with datasource=[%s] and interval=[%s] for task [%s] with priority=[%s] groupId=[%s]",
+          entry.getValue().getDataSource(),
+          entry.getValue().getInterval(),
+          task.getId(),
+          task.getPriority(),
+          task.getGroupId()
+      );
+      unlock(task, entry.getValue().getInterval());
+    }
+  }
+
+  /**
+   * Given GroupId, DataSource, Interval, Version and Priority create a new TaskLockPosse
+   * Add the TaskLockPosse to the in-memory data structure keeping track of all the TaskLockPosse sorted by interval for a datasource
+   * */
+  private TaskLockPosse createNewTaskLockPosse(String groupId, String dataSource, Interval interval, String version, Integer priority) {
+    if (!running.containsKey(dataSource)) {
+      running.put(dataSource, new TreeMap<Interval, TaskLockPosse>(Comparators.intervalsByStartThenEnd()));
+    }
+    TaskLockPosse taskLockPosse = new TaskLockPosse(new TaskLock(groupId, dataSource, interval, version, priority, true));
+    running.get(dataSource).put(interval, taskLockPosse);
+    log.info("Created new TaskLockPosse: %s", taskLockPosse);
+    return taskLockPosse;
+  }
+
+  /**
+   * Upgrade the taskLock for this task to non-preemptive lock
+   */
+  public boolean upgradeLock (Task task, Interval interval) {
+    giant.lock();
+    final String taskId = task.getId();
+    try {
+      // Upgrade the TaskLocks associated with this Task to non-preemptive lock
+      // If no higher priority task has arrived then there should still be a valid lock present inside TaskLockbox
+      List<TaskLockPosse> taskLockPosses = findLockPossesForTask(task);
+      List<TaskLockPosse> taskLockPossesToUpdgrade = new ArrayList<>();
+      for (TaskLockPosse taskLockPosse : taskLockPosses) {
+        // Second check is not really required as a task operates on a single datasource but anyways
+        if (taskLockPosse.getTaskLock().getInterval().equals(interval) &&
+            taskLockPosse.getTaskLock().getDataSource().equals(task.getDataSource())){
+          taskLockPossesToUpdgrade.add(taskLockPosse);
+        }
+      }
+      // There should only be one TaskLockPosse corresponding to a task and an interval at TaskLockbox
+      TaskLockPosse taskLockPosseToUpgrade = Iterables.getOnlyElement(taskLockPossesToUpdgrade);
+      TaskLock nonPreemptiveTaskLock = taskLockPosseToUpgrade.getTaskLock().withPreemptive(false);
+      // Upgrade the lock to non-preemptive lock in the TaskStorage
+      // Side-effect - this will cause all the tasks in the TaskLockPosse to have non-preemptive lock although they may not want it
+      // This should not be a problem as of now as every task has unique group Id and thus no two tasks share a TaskLockPosse
```

Comment:
No. It should not break the groupId. The behavior with groupId should not change with this PR. It is just that all the tasks sharing this task lock will have the ability to publish segment after any one of them gets non-preemptive lock which is I guess OK, as they all all sharing the lock for same datasource and interval. This commonly happens when tranquility runs replicated tasks, they all have same groupId and thus ability to publish same segment. 
I was a bit skeptical about all of the task sharing same task lock to get non-preemptive privileges but now I think it should be OK and I should remove the comment.